admin/services/redis-cache.service.js

```diff
@@ -0,0 +1,215 @@
+  const keyParts = serviceKey.split('.')
+  if (keyParts.length > 2) {
+    // the key includes an identifier e.g. school_id
+    affectedKey = `${keyParts[0]}.${keyParts[1]}`
+  }
+  const tables = redisCacheService.affectedTables[affectedKey]
+  if (tables) {
+    serviceKey = `${serviceKey}_${tables.join('-')}`
+  }
+  return serviceKey
+}
+
+/**
+ * Returns the data from a Redis key entry
+ * @param redisKey - the full Redis key to get
+ * @returns {String}
+ */
+redisCacheService.get = redisKey => {
+  return new Promise((resolve, reject) => {
+    redis.get(redisKey, (err, result) => {
+      if (err) {
+        logger.error(`REDIS (get): Error: ${err.message}`)
+        reject(err)
+      } else if (!result) {
+        logger.info(`REDIS (get): \`${redisKey}\` is not set`)
+        resolve(false)
+      } else {
+        logger.info(`REDIS (get): Retrieved \`${redisKey}\``)
+        resolve(result)
+      }
+    })
+  })
+}
+
+/**
+ * Stores data for a Redis key
+ * @param redisKey - the full Redis key to set
+ * @param data - string data to insert
+ * @returns {Boolean}
+ */
+redisCacheService.set = (redisKey, data) => {
+  return new Promise((resolve, reject) => {
+    if (typeof data === 'object') {
+      data = JSON.stringify(data)
+    }
+    redis.set(redisKey, data, () => {
+      logger.info(`REDIS (set): Stored \`${redisKey}\``)
+      resolve(true)
+    })
+  })
+}
+
+/**
+ * Called by the sqlService when it performs queries.
+ * If a change query is detected, finds table names and drops any affected caches so they can be redone
+ * @param sql - The SQL query being executed
+ * @returns {Boolean}
+ */
+redisCacheService.dropAffectedCaches = sql => {
+  return new Promise((resolve, reject) => {
+    if (!REDIS_CACHING) {
+      return resolve(true)
+    }
+    let tables = []
+    if (/DELETE|INSERT|UPDATE/.test(sql)) {
+      const tableRegex = /\.\[([a-z]+)\]/gi
+      let match = false
+      while ((match = tableRegex.exec(sql)) !== null) {
+        tables.push(match[1])
+      }
+    }
+    if (/EXEC/.test(sql)) {
+      const spRegex = /\.\[(sp[a-z]+)\]/gi
+      let match = false
+      let procedures = []
+      while ((match = spRegex.exec(sql)) !== null) {
+        procedures.push(match[1])
+      }
+      procedures.forEach(p => {
+        const thisTables = redisCacheService.storedProceduresAffectedTables[p]
+        if (thisTables) {
+          tables = tables.concat(thisTables)
+        }
+      })
+    }
+    if (tables.length) {
+      const stream = redis.scanStream()
+      stream.on('data', keys => {
+        const keyRegex = new RegExp(`(_|-)(${tables.join('|')})`, 'i')
+        const matchedKeys = keys.filter(k => keyRegex.test(k))
+        if (matchedKeys.length) {
+          const pipeline = redis.pipeline()
+          matchedKeys.forEach(key => {
+            logger.info(`REDIS (dropAffectedCaches): Dropped \`${key}\``)
+            pipeline.del(key)
+          })
+          pipeline.exec()
```

Comment:
should we be awaiting this?

Reply:
The structure has changed, but it is now an await.